### PR TITLE
Add HTTP Middleware

### PR DESCRIPTION
**Pull request by toma3233**
Added panic handling and logging middleware specifically for HTTP requests.

### CODE REVIEW

File: httpmw/httpmw_test.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+			Expect(buf.String()).To(ContainSubstring("finished call"))
+			Expect(rw.Result().StatusCode).To(Equal(200))
+
+		})
+	})
+})
+
+// Custom panic handler for testing
+func customPanicHandler(logger *slog.Logger, w http.ResponseWriter, r *http.Request, err interface{}) {
```
Comment (NickKeller) [inline httpmw/httpmw_test.go:69]: can we use a logging interface rather than a specific logging struct? This way users can plumb in whichever logging source they want to use

File: httpmw/logging_mw.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+	customWriter := &responseWriter{ResponseWriter: w}
+
+	startTime := l.now()
+	l.next.ServeHTTP(customWriter, r)
+	endTime := l.now()
+
+	latency := endTime.Sub(startTime)
+
+	logger.With(
```
Comment (NickKeller) [inline httpmw/logging_mw.go:57]: Can you use an interface method like `logRequestStart` and `logRequestEnd` to specify that the call is started/finished respectively? Different services log requests differently, and using an interface method would allow users to plumb in whatever implementation they desire.

From the aks-rp side, we use the messages "***RequestStart" and "***RequestEnd" to designate them, and a lot of our alerts filter off of those messages. Using an interface would allow basically a drop in replacement, without the need to go and change any other alerts

**Review by NickKeller — APPROVED** [review]

File: httpmw/httpmw_test.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+			logger := slog.New(slog.NewJSONHandler(os.Stdout, nil))
+			router.Use(NewPanicHandling(logger, customPanicHandler))
+			router.HandleFunc("/", func(w http.ResponseWriter, e *http.Request) {
+				panic("oops")
+			})
+
+			rw := httptest.NewRecorder()
+			req := httptest.NewRequest("GET", "/", nil)
+
+			router.ServeHTTP(rw, req)
```
Comment (fangluguomsft) [inline httpmw/httpmw_test.go:37]: This is not the typical way of test http middleware.

Examples:

 ```
    // Create a new router and apply the middleware.
    r := mux.NewRouter()
    r.Use(LoggingMiddleware)
    r.Handle("/", testHandler)

    // Create a test HTTP server.
    ts := httptest.NewServer(r)
    defer ts.Close()

    // Make a request to the test server.
    resp, err := http.Get(ts.URL)
    if err != nil {
        t.Fatalf("Failed to make GET request: %v", err)
    }
    defer resp.Body.Close()

    // Check the response status code.
    if resp.StatusCode != http.StatusOK {
        t.Errorf("Expected status code %d, got %d", http.StatusOK, resp.StatusCode)
    }

    // Check the response body.
    body := new(strings.Builder)
    _, err = body.ReadFrom(resp.Body)
    if err != nil {
        t.Fatalf("Failed to read response body: %v", err)
    }
    if body.String() != "Test passed" {
        t.Errorf("Expected response body 'Test passed', got '%s'", body.String())
    }
```

File: httpmw/logging_mw.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	now    func() time.Time
+	logger Logger
+}
+
+type responseWriter struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+func (rw *responseWriter) WriteHeader(code int) {
```
Comment (fangluguomsft) [inline httpmw/logging_mw.go:38]: rw typically written as w. rw make me feel it is read write.

File: httpmw/logging_mw.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+func (rw *responseWriter) Write(b []byte) (int, error) {
+	if rw.statusCode == 0 {
+		rw.statusCode = http.StatusOK
+	}
+	return rw.ResponseWriter.Write(b)
+}
+
+func (l *loggingMiddleware) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	customWriter := &responseWriter{ResponseWriter: w}
```
Comment (fangluguomsft) [inline httpmw/logging_mw.go:51]: I think it probably is not hard to get the statusCode from w. Why bother to create responseWriter{}?

File: httpmw/logging_mw.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+	"github.com/gorilla/mux"
+)
+
+type Logger interface {
+	Info(msg string, keysAndValues ...interface{})
+	Error(msg string, keysAndValues ...interface{})
+}
+
+func NewLogging(logger Logger) mux.MiddlewareFunc {
```
Comment (fangluguomsft) [inline httpmw/logging_mw.go:15]: I didn't find examples how this part works. If you have links, please add as comment.

File: httpmw/logging_mw.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+	startTime := l.now()
+	l.LogRequestStart(r)
+	l.next.ServeHTTP(customWriter, r)
+	endTime := l.now()
+
+	latency := endTime.Sub(startTime)
+	l.LogRequestEnd(r, customWriter.statusCode, latency)
+
+	l.logger.Info("finished call", "source", "ApiRequestLog", "protocol", "HTTP", "method_type", "unary", "component", "client", "time_ms", latency.Milliseconds(), "method", r.Method, "service", r.Host, "url", r.URL.String(), "code", customWriter.statusCode)
```
[the remark this answers is not in the document]
Comment (toma3233) [reply httpmw/logging_mw.go:61]: Reason I added both is because several alerts in aks-rp utilize "RequestStart" and "RequestEnd" messages, and our own service hub dashboards filter for "finished call". 

Added "msg" parameter to LogRequestEnd so that we can use for it for both cases

File: httpmw/httpmw_test.go (hunk 3 above)
Comment (toma3233) [reply httpmw/httpmw_test.go:37]: Followed the standard set by Fleet team for this test: https://msazure.visualstudio.com/CloudNativeCompute/_git/aks-rp?path=/fleet/pkg/api/middleware/panic_handling_mw_test.go&version=GBmaster&_a=contents

Can adjust test if needed.

File: httpmw/logging_mw.go (hunk 6 above)
Comment (toma3233) [reply httpmw/logging_mw.go:15]: The mux.MiddlewareFunc is a type defined in gorilla/mux which is used to wrap the existing http handler to add logging functionality in this case. 

Here is an example of usage in Fleet: https://msazure.visualstudio.com/CloudNativeCompute/_git/aks-rp?path=/fleet/pkg/api/middleware/logging_mw.go&version=GBmaster&line=21&lineEnd=22&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

Added comment in the code with link to gorilla/mux doc.

File: httpmw/httpmw_suite_test.go [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package httpmw
```
[the remark this answers is not in the document]
Comment (toma3233) [reply httpmw/httpmw_suite_test.go:1]: done

File: httpmw/logging_mw.go (hunk 5 above)
Comment (toma3233) [reply httpmw/logging_mw.go:51]: The http.ResponseWriter interface does not expose the status code for us to easily access. 

https://stackoverflow.com/questions/53272536/how-do-i-get-response-statuscode-in-golang-middleware

File: httpmw/logging_mw.go (hunk 4 above)
Comment (toma3233) [reply httpmw/logging_mw.go:38]: Oops, yeah that is supposed to be w. Made the change.

File: httpmw/panic_recovery_mw.go [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+type PanicHandlerFunc func(logger Logger, w http.ResponseWriter, r *http.Request, err interface{})
+
+func NewPanicHandling(logger Logger, panicHandler PanicHandlerFunc) mux.MiddlewareFunc {
+	return func(next http.Handler) http.Handler {
+		return &panicHandlingMiddleware{
+			next:         next,
+			logger:       logger,
+			panicHandler: panicHandler,
```
[the remark this answers is not in the document]
Comment (toma3233) [reply httpmw/panic_recovery_mw.go:16]: Added default handler and unit test to ensure default handler is used if no custom handler is passed in

File: httpmw/panic_recovery_mw.go [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+	}
+}
+
+type panicHandlingMiddleware struct {
+	next         http.Handler
+	logger       Logger
+	panicHandler PanicHandlerFunc
+}
+
+func (op *panicHandlingMiddleware) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```
[the remark this answers is not in the document]
Comment (toma3233) [reply httpmw/panic_recovery_mw.go:27]: switched to just p

File: httpmw/httpmw_test.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+	"net/http/httptest"
+	"os"
+	"time"
+
+	"github.com/gorilla/mux"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Httpmw", func() {
```
[the remark this answers is not in the document]
Comment (toma3233) [reply httpmw/httpmw_test.go]: done

File: httpmw/logging/logging_test.go [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+		router = mux.NewRouter()
+	})
+
+	Describe("LoggingMiddleware", func() {
+		It("should log and return OK status", func() {
+			buf := new(bytes.Buffer)
+			slogLogger := slog.New(slog.NewJSONHandler(buf, nil))
+			router.Use(NewLogging(slogLogger))
+
+			router.HandleFunc("/", func(w http.ResponseWriter, e *http.Request) {
```
Comment (fangluguomsft) [inline httpmw/logging/logging_test.go:30]: e *http.Request

maybe use _ to show it is not used.

File: httpmw/logging/logging_test.go [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+				time.Sleep(10 * time.Millisecond)
+				w.WriteHeader(http.StatusOK)
+			})
+
+			rw := httptest.NewRecorder()
```
Comment (fangluguomsft) [inline httpmw/logging/logging_test.go:35]: rw

Be consistent with other code. Stick with var name w.

File: httpmw/logging/logging.go [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+	Error(msg string, keysAndValues ...interface{})
+}
+
+// more info about gorilla/mux mw here:
+// https://github.com/gorilla/mux/tree/db9d1d0073d27a0a2d9a8c1bc52aa0af4374d265?tab=readme-ov-file#middleware
+func NewLogging(logger Logger) mux.MiddlewareFunc {
+	return func(next http.Handler) http.Handler {
+		return &loggingMiddleware{
+			next:   next,
+			now:    time.Now,
```
Comment (fangluguomsft) [inline httpmw/logging/logging.go:21]: Why register the function? It doesn't seem necessary.

File: httpmw/logging/logging.go [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+	"github.com/gorilla/mux"
+)
+
+type Logger interface {
+	Info(msg string, keysAndValues ...interface{})
+	Error(msg string, keysAndValues ...interface{})
+}
+
+// more info about gorilla/mux mw here:
+// https://github.com/gorilla/mux/tree/db9d1d0073d27a0a2d9a8c1bc52aa0af4374d265?tab=readme-ov-file#middleware
```
Comment (fangluguomsft) [inline httpmw/logging/logging.go:16]: This is a better link: https://pkg.go.dev/net/http#Handler

File: httpmw/recovery/recovery.go [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+type panicHandlingMiddleware struct {
+	next         http.Handler
+	logger       logging.Logger
+	panicHandler PanicHandlerFunc
+}
+
+func (p *panicHandlingMiddleware) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer func() {
```
Comment (fangluguomsft) [inline httpmw/recovery/recovery.go:37]: Any difference between defer vs just run the code after the line of p.next.ServeHTTP(w, r)?

File: httpmw/recovery/recovery_test.go [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+			router.HandleFunc("/", func(w http.ResponseWriter, e *http.Request) {
+				panic("oops")
+			})
+
+			rw := httptest.NewRecorder()
+			req := httptest.NewRequest("GET", "/", nil)
+
+			router.ServeHTTP(rw, req)
+
+			Expect(rw.Body.String()).To(ContainSubstring("Internal Server Error"))
```
Comment (fangluguomsft) [inline httpmw/recovery/recovery_test.go:38]: Change the expectation to differentiate it from the default handler.

File: httpmw/logging/logging.go [hunk 18]
```diff
@@ -0,0 +1,71 @@
+package logging
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/gorilla/mux"
+)
+
+type Logger interface {
```
Comment (fangluguomsft) [inline httpmw/logging/logging.go:10]: Do it in another PR.

Either move this to a different package that can be used by the gPRC middleware
Or just use the slogger interface if it is already there. Arbitrarily defining two functions doesn't seem a well thought interface.

File: httpmw/logging/logging.go [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+	return func(next http.Handler) http.Handler {
+		return &loggingMiddleware{
+			next:   next,
+			now:    time.Now,
+			logger: logger,
+		}
+	}
+}
+
+var _ http.Handler = &loggingMiddleware{}
```
[the remark this answers is not in the document]
Comment (toma3233) [reply httpmw/logging/logging.go:27]: Yes, here I am enforcing that loggingMiddleware implements the http.Handler interface to ensure safety at compile time. This is an example of something that fleet does that I decided to incorporate into our middleware. 

https://msazure.visualstudio.com/CloudNativeCompute/_git/aks-rp?path=/fleet/pkg/api/middleware/logging_mw.go&version=GBmaster&line=37&lineEnd=38&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

File: httpmw/logging/logging.go [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+	l.LogRequestEnd(r, "RequestEnd", customWriter.statusCode, latency)
+	l.LogRequestEnd(r, "finished call", customWriter.statusCode, latency)
+}
+
+func (l *loggingMiddleware) LogRequestStart(r *http.Request) {
+	l.logger.Info("RequestStart", "source", "ApiRequestLog", "protocol", "HTTP", "method_type", "unary", "component", "client", "method", r.Method, "service", r.Host, "url", r.URL.String())
+}
+
+func (l *loggingMiddleware) LogRequestEnd(r *http.Request, msg string, statusCode int, duration time.Duration) {
+	l.logger.Info(msg, "source", "ApiRequestLog", "protocol", "HTTP", "method_type", "unary", "component", "client", "method", r.Method, "service", r.Host, "url", r.URL.String(), "code", statusCode, "time_ms", duration.Milliseconds())
```
[the remark this answers is not in the document]
Comment (toma3233) [reply httpmw/logging/logging.go:70]: done

File: httpmw/logging/logging_test.go (hunk 12 above)
Comment (toma3233) [reply httpmw/logging/logging_test.go:30]: done

File: httpmw/logging/logging_test.go (hunk 13 above)
Comment (toma3233) [reply httpmw/logging/logging_test.go:35]: done

File: httpmw/logging/logging.go (hunk 14 above)
Comment (toma3233) [reply httpmw/logging/logging.go:21]: By returning func(http.Handler) http.Handler, we are adhering to the mux.Middleware func which is needed to easily register our middleware to a router. The NewLogging function encapsulates the logging fucntionality and allows us to use it with any logger we pass in. 

image

https://github.com/gorilla/mux/tree/db9d1d0073d27a0a2d9a8c1bc52aa0af4374d265?tab=readme-ov-file#middleware

File: httpmw/recovery/recovery.go (hunk 16 above)
Comment (toma3233) [reply httpmw/recovery/recovery.go:37]: Without defer, If p.next.ServeHTTP(w, r) or any of the other calls in the chain panics, the execution will stop and the recovery logic will not be effective. 

With the defer, the process will defer to the recovery logic when a panic is encountered, resulting in the panic being handled gracefully.

https://www.jajaldoang.com/post/handle-panic-in-http-server-using-middleware-go/

File: httpmw/recovery/recovery_test.go (hunk 17 above)
Comment (toma3233) [reply httpmw/recovery/recovery_test.go:38]: done

File: httpmw/logging/logging.go (hunk 15 above)
Comment (toma3233) [reply httpmw/logging/logging.go:16]: Oh yes, that is good for info about the http handler type. Added.

File: httpmw/logging/logging.go (hunk 18 above)
Comment (toma3233) [reply httpmw/logging/logging.go]: Added a todo to add in another PR. Will, implement something similar to the article below.

https://medium.com/@ansujain/building-a-logger-wrapper-in-go-with-support-for-multiple-logging-libraries-48092b826bee

File: httpmw/logging/logging.go (hunk 14 above)
Comment (fangluguomsft) [reply httpmw/logging/logging.go:19]: I mean why use field now to register the time.Now() function?

File: httpmw/logging/logging.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (fangluguomsft) [reply httpmw/logging/logging.go]: Still quite long. You can add new line after comma.

**Review by fangluguomsft — APPROVED** [review]

File: httpmw/logging/logging.go (hunk 14 above)
Comment (toma3233) [reply httpmw/logging/logging.go:19]: It allows for better testing since we can mock it, and control the time to test certain behaviors. It also ensures consistency so that the same time function is used everywhere in the middleware. 

The Fleet logging mw utilizes this as well.

File: httpmw/logging/logging.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (toma3233) [reply httpmw/logging/logging.go]: done
